src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,80 @@
+package duke;
+
+import duke.command.Command;
+import duke.exception.DukeException;
+
+public class Duke {
+
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
+
+    /**
+     * Initialise Duke with stated filePath.
+     *
+     * @param filePath filePath to read and save list data.
+     */
+    public Duke(String filePath) {
+        ui = new Ui();
+        storage = new Storage(filePath);
+        try {
+            tasks = storage.load();
+        } catch (DukeException e) {
+            ui.show(e.getMessage());
+            tasks = new TaskList();
+        }
+    }
+
+
+    public Duke usingMainUi() {
+        this.ui = new MainUi();
+        return this;
+    }
+
+    public static void main(String[] args) {
+        new Duke("data/duke.ser").run();
+    }
+
+    /**
+     * Main loop of Duke.
+     */
+    public void run() {
+        assert tasks != null;
+        assert ui != null;
+        assert storage != null;
+
+        ui.showWelcome();
+        boolean isExit = false;
+        while (!isExit) {
+            try {
+                Command c = Parser.parse(ui.readCommand());
+                c.execute(tasks, ui, storage);
+                isExit = c.isExit();
+            } catch (DukeException e) {
+                ui.show(e.getMessage());
+            }
+        }
+    }
+
+
+    /**
+     * Tries to parse and execute input as Command to Duke.
+     * Any exceptions should be included within String response.
+     *
+     * @param input Parses and Executes input as Command to Duke
+     * @return Duke's response as String
+     */
+    public String getResponse(String input) {
+        assert input != null;
+        assert ui instanceof MainUi;
+
+        try {
+            Command c = Parser.parse(input);
+            c.execute(tasks, ui, storage);
+        } catch (DukeException e) {
+            ui.show(e.getMessage());
+        } finally {
+            return ((MainUi) ui).getResponse();
```

Comment:
Why did you need to typecast your Ui?

Reply:
MainWindow uses MainUi while Ui doesn't have any definition of getResponse(). There is alternative ways of cleaner implementation but I believe but I thought it would be simpler to cast it. Might put this on forum for discussion on good practice for non-existent subclass methods in super-class.